### PR TITLE
Merge with image top tag producer

### PR DESCRIPTION
**Pull request by knash**
First working version of image top producer.  
The training file is preliminary and outdated so there will need to be an additional update - But the changes will likely be mostly to the training file alone, so it is good to merge the producer portion now to test compatibility.

### CODE REVIEW

**Comment by knash** [comment]
Hello, what was the issue with this merge?

**Comment by knash** [comment]
Update to final training -- now is a good time to review and merge

File: plugins/ImageProducer.cc [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+#include "DataFormats/PatCandidates/interface/PackedCandidate.h"
+
+#include "DataFormats/PatCandidates/interface/Jet.h"
+
+
+#include "tensorflow/c/c_api.h"
```
Comment (hqucms) [inline plugins/ImageProducer.cc]: This can probably be removed.

File: plugins/ImageProducer.cc [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+#include "PhysicsTools/TensorFlow/interface/TensorFlow.h"
+#include "TLorentzVector.h"
+
+#include <iostream>
+#include <fstream>
+
+#include <memory>
+#include <iostream>
+#include <Python.h>
```
Comment (hqucms) [inline plugins/ImageProducer.cc]:
```c++
#include <Python.h>
```
Not used I think?

File: plugins/ImageProducer.cc [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+  //From deepflavour implementation, for consistency 
+  std::atomic<tensorflow::GraphDef*> graphDef;
+};
+
+class ImageProducer : public edm::stream::EDProducer<edm::GlobalCache<ImageTFCache>> {
+
+  public:
+    explicit ImageProducer(const edm::ParameterSet&, const ImageTFCache*);
+    ~ImageProducer() override;
+    double principal_axis(std::vector<std::vector<float>>);
```
Comment (hqucms) [inline plugins/ImageProducer.cc]: `const std::vector<std::vector<float>> &`

File: plugins/ImageProducer.cc [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+    static void fillDescriptions(edm::ConfigurationDescriptions&);
+    static void globalEndJob(const ImageTFCache*);
+    static std::unique_ptr<ImageTFCache> initializeGlobalCache(const edm::ParameterSet&);
+    std::uniform_real_distribution<> dis = std::uniform_real_distribution<>(0.0,1.0);
```
Comment (hqucms) [inline plugins/ImageProducer.cc]: Not used?

File: plugins/ImageProducer.cc [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+
+    //just for debug text3
+    std::ofstream textout;
```
Comment (hqucms) [inline plugins/ImageProducer.cc]: remove?

File: plugins/ImageProducer.cc [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+{
+  
+  vtx = consumes<reco::VertexCollection>(edm::InputTag("offlineSlimmedPrimaryVertices"));
+  
+  produces<pat::JetCollection>();
+
+  tensorflow::SessionOptions sessionOptions;
+  tfsession_ = tensorflow::createSession(cache->graphDef,sessionOptions);
+
+  if(iConfig.getParameter<edm::InputTag>("src").label()=="slimmedJetsAK8")sdmcoll="ak8PFJetsCHSSoftDropMass";
```
Comment (hqucms) [inline plugins/ImageProducer.cc]: Can you make this configurable in the python cfg file instead of hard-coded?

File: plugins/ImageProducer.cc [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+
+}
+ImageProducer::~ImageProducer(){
+ tensorflow::closeSession(tfsession_);
+}
+
+
+
+template <typename T> std::string to_string_pr(const T a_value, const int n = 17)
```
Comment (hqucms) [inline plugins/ImageProducer.cc]: Not used?

File: plugins/ImageProducer.cc [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+			else partlist[7].push_back(0.0);
+			idaufill+=1;
+			}
+		}
+
+  	for (const auto &subjet : *subjets)
+		{
+	       
+		sublv.SetPtEtaPhiM(subjet.pt(),subjet.eta(),subjet.phi(),subjet.mass());
+		if (sublv.DeltaR(curtlv)>0.8 || sjlist.size()>=12) continue;
```
Comment (hqucms) [inline plugins/ImageProducer.cc:200]: Why this restriction with `sjlist.size()>=12`?

File: python/ak8_cff.py [hunk 9]
```diff
@@ -124,6 +151,7 @@ def setupCustomizedAK8(process, runOnMC=False, path=None):
             bestH=Var("userFloat('BEST:dnn_higgs')", float, doc="Boosted Event Shape Tagger score Higgs", precision=-1),
             bestQCD=Var("userFloat('BEST:dnn_qcd')", float, doc="Boosted Event Shape Tagger score QCD", precision=-1),
             bestB=Var("userFloat('BEST:dnn_b')", float, doc="Boosted Event Shape Tagger score B", precision=-1),
+            itop=Var("userFloat('Image:top')", float, doc="Image tagger score top", precision=-1),
```
Comment (hqucms) [inline python/ak8_cff.py:151]: Can you rename the variable to `imageTop` to be more self-explanatory?

File: BuildFile.xml [hunk 10]
```diff
@@ -5,6 +5,10 @@
 <use   name="PhysicsTools/CandUtils"/>
 <use   name="fastjet"/>
 <use   name="lwtnn"/>
+<use   name="tensorflow-cc" />
```
Comment (hqucms) [inline BuildFile.xml]: This can probably be removed.

File: plugins/MuSubProducer.cc [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+#include "TLorentzVector.h"
+
+#include <iostream>
+#include <fstream>
+
+#include <memory>
+#include <iostream>
+#include <Python.h>
+
```
Comment (hqucms) [inline plugins/MuSubProducer.cc]:
```c++
#include "TLorentzVector.h"

#include <iostream>
#include <fstream>

#include <memory>
#include <iostream>
#include <Python.h>
```

I think these are not needed?

File: plugins/MuSubProducer.cc [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+  iEvent.getByToken(vtxs_, vtxs);
+  iEvent.getByToken(src_, muons);
+  iEvent.getByToken(pfcs_, PackedCandidates);
+  for(const auto &cand : *PackedCandidates)   
+	{
+	bool keepcand=true;
+	for (const auto &curmuon : *muons)
+		{
+		if((curmuon.pt()>ptmin_) and (curmuon.isTightMuon(vtxs->at(0))))
```
Comment (hqucms) [inline plugins/MuSubProducer.cc]: Probably need to add a protection for empty `vtxs` collection?

File: python/ak4_cff.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        'pfDeepCSVJetTags:probbb'
+    ]
+    JETCorrLevels = ['L1FastJet', 'L2Relative', 'L3Absolute', 'L2L3Residual']
+
+    from PhysicsTools.NanoHRT.jetToolbox_cff import jetToolbox
+    jetToolbox(process, 'ak4', 'dummySeq', 'out', associateTask=False,
+               PUMethod='CHS', JETCorrPayload='AK4PFchs', JETCorrLevels=JETCorrLevels,
+	       postFix='MuSub',	
+	       newPFCollection=True, nameNewPFCollection="MuSubProducer",
+               Cut='pt > 30.0 && abs(rapidity()) < 2.4',
```
Comment (hqucms) [inline python/ak4_cff.py]: I think this `Cut` is before applying the JEC, so maybe we need to leave some margin (e.g., pt>20 or 25)?

File: plugins/ImageProducer.cc (hunk 1 above)
Comment (knash) [reply plugins/ImageProducer.cc]: jet is required but not tensorflow it seems

File: plugins/ImageProducer.cc (hunk 2 above)
Comment (knash) [reply plugins/ImageProducer.cc]: done

File: plugins/ImageProducer.cc (hunk 3 above)
Comment (knash) [reply plugins/ImageProducer.cc]: done

File: plugins/ImageProducer.cc (hunk 4 above)
Comment (knash) [reply plugins/ImageProducer.cc]: Correct, but this was from an earlier PR?  It is not in the latest version

File: plugins/ImageProducer.cc (hunk 5 above)
Comment (knash) [reply plugins/ImageProducer.cc]: Also not in latest version

File: plugins/ImageProducer.cc (hunk 6 above)
Comment (knash) [reply plugins/ImageProducer.cc]: done

File: plugins/ImageProducer.cc (hunk 7 above)
Comment (knash) [reply plugins/ImageProducer.cc]: done

File: plugins/ImageProducer.cc (hunk 8 above)
Comment (knash) [reply plugins/ImageProducer.cc:200]: sjlist.size()==12 is when two subjets are found. 
The way we match means there is a very rare case where three are associated depending on the input subjet collection (which we want to keep general for now), but we train with at most two so these are skipped.

File: python/ak8_cff.py (hunk 9 above)
Comment (knash) [reply python/ak8_cff.py:151]: Would prefer to keep as-is.  There will be mutiple nodes in the next version

File: BuildFile.xml (hunk 10 above)
Comment (knash) [reply BuildFile.xml]: done.  IIRC this was needed at some point but not so much now it seems

File: plugins/MuSubProducer.cc (hunk 11 above)
Comment (knash) [reply plugins/MuSubProducer.cc]: done

File: plugins/MuSubProducer.cc (hunk 12 above)
Comment (knash) [reply plugins/MuSubProducer.cc]: done

File: python/ak4_cff.py (hunk 13 above)
Comment (knash) [reply python/ak4_cff.py]: done

**Comment by hqucms** [comment]
Merged w/ a few fixes https://github.com/hqucms/NanoHRT/commit/5e931e538ba71983bb14b046c35417cec951037a.

**Pull request merged**